cmd/algorand-indexer/daemon.go

```diff
@@ -61,253 +66,299 @@ var (
-
-	daemonCmd.Flags().StringVar(&suppliedAPIConfigFile, "api-config-file", "", "supply an API config file to enable/disable parameters")
-	daemonCmd.Flags().BoolVar(&enableAllParameters, "enable-all-parameters", false, "override default configuration and enable all parameters. Can't be used with --api-config-file")
-	daemonCmd.Flags().Uint32VarP(&maxAPIResourcesPerAccount, "max-api-resources-per-account", "", 1000, "set the maximum total number of resources (created assets, created apps, asset holdings, and application local state) per account that will be allowed in REST API lookupAccountByID and searchForAccounts responses before returning a 400 Bad Request. Set zero for no limit")
-	daemonCmd.Flags().Uint32VarP(&maxTransactionsLimit, "max-transactions-limit", "", 10000, "set the maximum allowed Limit parameter for querying transactions")
-	daemonCmd.Flags().Uint32VarP(&defaultTransactionsLimit, "default-transactions-limit", "", 1000, "set the default Limit parameter for querying transactions, if none is provided")
-	daemonCmd.Flags().Uint32VarP(&maxAccountsLimit, "max-accounts-limit", "", 1000, "set the maximum allowed Limit parameter for querying accounts")
-	daemonCmd.Flags().Uint32VarP(&defaultAccountsLimit, "default-accounts-limit", "", 100, "set the default Limit parameter for querying accounts, if none is provided")
-	daemonCmd.Flags().Uint32VarP(&maxAssetsLimit, "max-assets-limit", "", 1000, "set the maximum allowed Limit parameter for querying assets")
-	daemonCmd.Flags().Uint32VarP(&defaultAssetsLimit, "default-assets-limit", "", 100, "set the default Limit parameter for querying assets, if none is provided")
-	daemonCmd.Flags().Uint32VarP(&maxBalancesLimit, "max-balances-limit", "", 10000, "set the maximum allowed Limit parameter for querying balances")
-	daemonCmd.Flags().Uint32VarP(&defaultBalancesLimit, "default-balances-limit", "", 1000, "set the default Limit parameter for querying balances, if none is provided")
-	daemonCmd.Flags().Uint32VarP(&maxApplicationsLimit, "max-applications-limit", "", 1000, "set the maximum allowed Limit parameter for querying applications")
-	daemonCmd.Flags().Uint32VarP(&defaultApplicationsLimit, "default-applications-limit", "", 100, "set the default Limit parameter for querying applications, if none is provided")
-
-	daemonCmd.Flags().StringVarP(&indexerDataDir, "data-dir", "i", "", "path to indexer data dir, or $INDEXER_DATA")
-
-	daemonCmd.Flags().StringVarP(&cpuProfile, "cpuprofile", "", "", "file to record cpu profile to")
-	daemonCmd.Flags().StringVarP(&pidFilePath, "pidfile", "", "", "file to write daemon's process id to")
-	daemonCmd.Flags().StringVarP(&configFile, "configfile", "c", "", "file path to configuration file (indexer.yml)")
+	daemonConfig := newDaemonConfig()
+	daemonCmd = newDaemonCmd(daemonConfig)
```

Comment:
I want to move everything away from being called in init functions. So I've updated this to just have the root command directly create the config and cmd. Let me know what you think.

Reply:
My only concern is over-correcting and adding too much abstraction. We're using cobra, no need to hide it.

This is the pattern I've adopted lately: https://github.com/algorand/indexer/blob/develop/cmd/block-generator/runner/runner.go
Since the command exists already, the consumer just needs to execute it: https://github.com/algorand/indexer/blob/develop/cmd/block-generator/main.go

Its worked pretty nicely:
https://github.com/algorand/node-ui/blob/master/cmd/tui/main.go#L43
https://github.com/algorand/go-algorand/blob/master/cmd/algokey/keyreg.go#L60

That said, I'm happy to learn a new trick if you have a better idea.

Reply:
Yeah it doesn't matter too much to me--the pattern you're showing just uses the init function to construct the commands whereas I've put it in separate functions. 

I did have a feeling that relying on init functions seemed fragile to me so I investigated...and it can be.

The problem is that we have a command hierarchy--root command which adds subcommands. So when we run the `rootCmd.AddCommand` function we need to be sure that the subcommand we're adding is initialized.
If we use a variable such as this to define a command:  
`cmd/algorand-indexer/z.go`
```
package main

var var1 string
var zCmd := &cobra.Command{
	Use:   "z",
	Short: "foo",
	Long:  "bar",
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Println(var1)
	},
}

func init() {
    importCmd.Flags().StringVarP(&var1, "var1", "", "", "sample flag")
}
```
The above will work perfectly when we call `rootCmd.AddCommand(zCmd)` in `main.go`'s init function because the `zCmd` variable is "global/static" in the package, it will always be initialized before any init functions are called.

However, if we have a top level command variable defined which is initialized in the init function, such as this:
`cmd/algorand-indexer/z.go`
```
var zCmd *cobra.Command

var (
	var1 string
)

func init() {
	zCmd = &cobra.Command{
		Use:   "z",
		Short: "foo",
		Long:  "bar",
		Run: func(cmd *cobra.Command, args []string) {
			fmt.Println(var1)
		},
	}
	zCmd.Flags().StringVarP(&var1, "var1", "", "", "sample")
}
```
The above program will compile, but when you run the command you will get a seg fault:
```
./cmd/algorand-indexer/algorand-indexer z
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x104dc16bc]

goroutine 1 [running]:
github.com/spf13/cobra.(*Command).AddCommand(0x10686dda0, {0x1400051fd40, 0x1, 0x1})
	/Users/eric/.asdf/installs/golang/1.17.9/packages/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:1149 +0x5c
main.init.3()
	/Users/eric/go/src/github.com/algorand/indexer/cmd/algorand-indexer/main.go:121 +0x404
```
This is because init functions within the same package are run in lexicographical order. It just so happens that all of the commands we currently have come before `main.go`, but `z.go` is run afterwards so the variable is never initialized.

The reason this does not happen in the `block-generator` example above is that init functions in imports are run depth first, and the root command for block-generator is in a different package than the subcommands which are then imported here https://github.com/algorand/indexer/blob/develop/cmd/block-generator/core/commands.go#L7 causing their init functions to run before the block generator's `commands.go` init function.


Sorry for how long this post became 😸 , I just got really interested in this after you brought it up.

Reply:
So I think the best solution is one of:
* separate subcommands into their own packages
* use a top-level variable to initialize commands outside the init functions
* use dedicated constructor functions

Reply:
Sounds good, thanks for the explanation.